--- ./lib/softoken/pkcs11c.c.unused	2024-08-05 17:56:10.705414972 +0200
+++ ./lib/softoken/pkcs11c.c	2024-08-05 18:09:04.681015706 +0200
@@ -165,10 +165,14 @@ SFTKCipherWrap(AESKeyWrapContext, AESKey
         mmm##_DestroyContext(ctx, freeit);                                   \
     }
 
+#ifndef NSS_DISABLE_DEPRECATED_RC2
 SFTKCipherWrap2(RC2Context, RC2);
+#endif
 SFTKCipherWrap2(RC4Context, RC4);
 SFTKCipherWrap2(DESContext, DES);
+#ifndef NSS_DISABLE_DEPRECATED_SEED
 SFTKCipherWrap2(SEEDContext, SEED);
+#endif
 SFTKCipherWrap2(CamelliaContext, Camellia);
 SFTKCipherWrap2(AESContext, AES);
 SFTKCipherWrap2(AESKeyWrapContext, AESKeyWrap);
