diff --git a/lib/softoken/fipstokn.c b/lib/softoken/fipstokn.c
--- a/lib/softoken/fipstokn.c
+++ b/lib/softoken/fipstokn.c
@@ -244,17 +244,17 @@ fc_getAttribute(CK_ATTRIBUTE_PTR pTempla
 
 #define CK_PKCS11_FUNCTION_INFO(name) CK_RV __PASTE(F, name)
 #define CK_NEED_ARG_LIST 1
 
 #include "pkcs11f.h"
 
 /* ------------- build the CK_CRYPTO_TABLE ------------------------- */
 static CK_FUNCTION_LIST_3_0 sftk_fipsTable = {
-    { CRYPTOKI_VERSION_MAJOR, CRYPTOKI_VERSION_MINOR },
+    { 3, 0 },
 
 #undef CK_NEED_ARG_LIST
 #undef CK_PKCS11_FUNCTION_INFO
 
 #define CK_PKCS11_FUNCTION_INFO(name) \
     __PASTE(F, name)                  \
     ,
 
diff --git a/lib/softoken/pkcs11.c b/lib/softoken/pkcs11.c
--- a/lib/softoken/pkcs11.c
+++ b/lib/softoken/pkcs11.c
@@ -107,17 +107,17 @@ static PRIntervalTime loginWaitTime;
 #define CK_PKCS11_FUNCTION_INFO(name) CK_RV __PASTE(F, name)
 #define CK_NEED_ARG_LIST 1
 
 #include "pkcs11f.h"
 #endif
 
 /* build the crypto module table */
 static CK_FUNCTION_LIST_3_0 sftk_funcList = {
-    { CRYPTOKI_VERSION_MAJOR, CRYPTOKI_VERSION_MINOR },
+    { 3, 0 },
 
 #undef CK_PKCS11_FUNCTION_INFO
 #undef CK_NEED_ARG_LIST
 
 #define CK_PKCS11_FUNCTION_INFO(func) \
     __PASTE(NS, func)                 \
     ,
 #include "pkcs11f.h"
